### ISSUE TITLE
Soil level clm surface input data for clm5.2.0 have missing values in large domains

### ISSUE THREAD

**Issue by maritsandstad** [ESCOMP/CTSM#2744]
### Brief summary of bug

Input surface data for clm on spectral element grid ne30 (inputdata/lnd/clm2/surfdata_esmf/ctsm5.2.0/surfdata_ne30np4.pg3_hist_1850_78pfts_c240216.nc) at least for clm5.2.0 have weird/missing values for parts of the grid.

Here is an example showing organic carbon summed over the soil levels and similar patterns of missing data seem to be present in all the soil level data I've tried plotting including PCT_CLAY, PCT_SAND:

![ORGC_output_se_toplevel_map](https://github.com/user-attachments/assets/e9779552-649f-4e5f-bff2-491445a118ba)

This pattern shows up similarly in output datasets

### General bug information

Plotting and regridding input data from surface input data for spectral element grid ne30 for clm5.2.0 (inputdata/lnd/clm2/surfdata_esmf/ctsm5.2.0/surfdata_ne30np4.pg3_hist_1850_78pfts_c240216.nc) data for variables on soil levels seem to be looking weird and missing data for what looks like some of the spectral element domains.

Above is an example plot of what organic carbon summed over soil levels looks like, but the pattern of missing data seems to be the same for all the soil level data that I have plotted so far. Looking at the LMR picture framing from https://journals.ametsoc.org/view/journals/mwre/134/12/mwr3360.1.xml I can almost see at least some of the domains that are missing, here is an attempt at colouring in the missing parts:

![Screenshot 2024-09-06 151006](https://github.com/user-attachments/assets/46330a44-2add-4f05-b72f-3d759c2ac956)

**CTSM version you are using:** [output of `git describe`]
clm5.2.0  ne30

**Does this bug cause significantly incorrect results in the model's science?** [Yes ]

**Configurations affected:** [Fill this in if known.]

### Details of bug

See above

### Important details of your setup / configuration so we can reproduce the bug

I regridded using the xesmf package following this excellent guide: https://ncar.github.io/esds/posts/2022/cam-se-regridding/

For reference here is my regridding code snippet
```
import numpy as np
import xarray as xr
import xesmf

def make_se_regridder(weight_file):
    weights = xr.open_dataset(weight_file)
    in_shape = weights.src_grid_dims.load().data

    # Since xESMF expects 2D vars, we'll insert a dummy dimension of size-1
    if len(in_shape) == 1:
        in_shape = [1, in_shape.item()]

    # output variable shape
    out_shape = weights.dst_grid_dims.load().data.tolist()[::-1]

    dummy_in = xr.Dataset(
        {
            "lat": ("lat", np.empty((in_shape[0],))),
            "lon": ("lon", np.empty((in_shape[1],))),
        }
    )
    dummy_out = xr.Dataset(
        {
            "lat": ("lat", weights.yc_b.data.reshape(out_shape)[:, 0]),
            "lon": ("lon", weights.xc_b.data.reshape(out_shape)[0, :]),
        }
    )

    regridder = xesmf.Regridder(
        dummy_in,
        dummy_out,
        weights=weight_file,
        method="bilinear",
        reuse_weights=True,
        periodic=True,
    )
    return regridder

def regrid_se_data(regridder, data_to_regrid):
    print(data_to_regrid.dims)
    if isinstance(data_to_regrid, xr.DataArray):
        print(type(data_to_regrid))
        updated = data_to_regrid.copy().transpose(..., "lndgrid").expand_dims("dummy", axis=-2)
    else:
        vars_with_ncol = [name for name in data_to_regrid.variables if "ncol" in data_to_regrid[name].dims]
        updated = data_to_regrid.copy().update(
            data_to_regrid[vars_with_ncol].transpose(..., "lndgrid").expand_dims("dummy", axis=-2)
        )
    regridded = regridder(updated.rename({"dummy": "lat", "lndgrid": "lon"}))
    return regridded
```

I used this file for mapping inputdata/lnd/clm2/mappingdata/maps/ne30pg3/map_ne30pg3_to_0.5x0.5_nomask_aave_da_c180515.nc

The regridding looked ok and completely sane for other variables that didn't have soil level, and the pattern persisted regardless of whether I regridded a variable with the soil level dimension, or if I summed over and thereby eliminated the soil level dimension before regridding. It also looked the same if I regridded the same output variables from the first history files from a run.

### Important output or errors that show the problem

See above

**Comment by maritsandstad** [comment]
I am happy to provide additional details or more plots. 

@mvertens, @mvdebolskiy @rosiealice, here is my write-up of the soil/dust data issue. Perhaps you also know whether this needs to flagged for anyone in particular.

**Comment by wwieder** [comment]
Thanks for raising this issue, @maritsandstad.

I wondered if this was an issue with the regridding or the dataset, so I tried plotting with uxarray, which seems to confirm your concern.  We'll have to look at this for the CTSM5.3 datasets that @slevis-lmwg and @ekluzek are working on finalizing now.

<img width="1126" alt="image" src="https://github.com/user-attachments/assets/89789428-9819-45a8-a808-4bdf30d3a733">

**Comment by wwieder** [comment]
visually, it looks like the same issue is present in the 5.3 surface data that @olyson is using for the ne30 spinup `/glade/derecho/scratch/slevis/temp_work/new_rawdata/tools/mksurfdata_esmf/surfdata_ne30np4.pg3_hist_1850_78pfts_c240905.nc`

**Comment by wwieder** [comment]
spot checking other data on the surface dataset it looks like this is just an issue for the soil variables, which I confirmed are messing up the soil properties being calculated by the pedotransfer functions in CLM (below for WATSAT.  
<img width="1124" alt="image" src="https://github.com/user-attachments/assets/a0c98d07-9750-410d-8316-dd5fbfa4c6be">

**Comment by wwieder** [comment]
I think this should be addressed in #2500 before we make the 5.3 tag.  It's also likely influencing dust emissions, @dmleung, assuming your using the ne30 grid in your testing?

**Comment by maritsandstad** [comment]
@wwieder, yeah I also just saw this for the soil variables, but I also saw the same for WATSAT from initial outputs from a testrun.

**Comment by mvertens** [comment]
@wwieder - can you look at any of the other unstructured surface datasets (e.g. ne30np4 - or any other refined grids) and see if the same problem arises. Is this just for ne30np4.pg3?

**Comment by billsacks** [comment]
It would also be very helpful to know if soil color shows this issue; has anyone checked that?

**Comment by samsrabin** [comment]
@wwieder notes that it "doesn't look like an issue in older surface datasets created by `mksurfdata_map` (`inputdata/lnd/clm2/surfdata_map/surfdata_ne30pg3_hist_16pfts_Irrig_CMIP6_simyr1850_c190227.nc`)"

**Comment by samsrabin** [comment]
> @wwieder - can you look at any of the other unstructured surface datasets (e.g. ne30np4 - or any other refined grids) and see if the same problem arises. Is this just for ne30np4.pg3?

Also worth checking to see if any structured-grid datasets are affected.

**Comment by billsacks** [comment]
I just spent a while on a call with @mvertens . In addition to the ideas noted above, we agreed that the best next step would be to output the regridded mapunit created as an intermediate step in creating these soil variables. I can make a branch this morning that adds that output.

**Comment by wwieder** [comment]
Good questions 
@billsacks soil color looks fine.

@samsrabin older datasets created with mksrfdata_map are OK 
(e.g. inputdata/lnd/clm2/surfdata_map/surfdata_ne30pg3_hist_16pfts_Irrig_CMIP6_simyr1850_c190227.nc)

@samsrabin a quick look at different structured grids at different resolutions look fine.

@mvertens other unstructured grids also look like the one originally noted here
ne30np4.pg2
ne3np4.pg3
The ne120 grid overloaded memory in jupyter hub.  More to come...

all the 5.3 datasets I'm looking at are here:
/glade/derecho/scratch/slevis/temp_work/new_rawdata/tools/mksurfdata_esmf/surfdata_ne30np4.pg2_hist_1850_78pfts_c240905.nc

**Comment by billsacks** [comment]
Thank you @wwieder - this is very helpful.

@ekluzek - @mvertens walked me through the regridding code for soil texture. The implementation is significantly different in mksurfdata_esmf compared with the old mksurfdata_map in some ways. But there are essentially two parts of what's going on: (1) regridding the mapunit and then (2) taking the regridded mapunit and from it creating and outputting the soil properties on the output grid. We felt that outputting the mapunit will let us narrow down the problem. It's especially interesting that soil color looks fine, because that uses a very similar (though not identical) algorithm to the regridding of mapunit for soil properties.

Anyway, I'll move ahead with adding the outputting of mapunit.

**Comment by wwieder** [comment]
This plan makes sense @billsacks checking that the mapunits are regridding correctly seems like the next step.

**Comment by wwieder** [comment]
FWIW the MPAS120 grid is NOT missing the regions in question (although plotting in uxarray look odd)

<img width="992" alt="image" src="https://github.com/user-attachments/assets/8dc79f1d-65f4-499e-a03d-f54aeb486ae5">

**Comment by wwieder** [comment]
The C96 grid is also affected

<img width="1112" alt="image" src="https://github.com/user-attachments/assets/e78336b3-71df-4a66-856b-587dc955bca2">

**Comment by mvertens** [comment]
@wwieder - thanks for looking at the other resolutions. It looks like the problem has to do with a cubed sphere grid (both C96 and spectral element resolutions are on a cubed sphere grid). So this analysis is really helpful.

**Comment by ekluzek** [comment]
@mvertens and @billsacks thanks so much for digging into this, we really appreciate it! This will take much less time for the two of you to solve.

After you are done it might be good to show more about this to the CTSM SE so we can also get some knowledge on this area.

And @maritsandstad thanks for pointing this out as well -- really helpful for the community to let us know about things like this.

**Comment by billsacks** [comment]
Oh, `mapunits` is already on the surface dataset. @wwieder can you please plot that for the ne30 surface dataset?

**Comment by dlawrenncar** [comment]
One other thing to think about in terms of implications for ongoing runs is whether or not this is affecting the km-scale simulations that @briandobbins is working on.  He is getting 'close' to a production-ish run and it would be good to know if this is impacting those super high-resolution runs.

**Comment by ekluzek** [comment]
@dlawrenncar this is looking like it's an issue with the cubed sphere grid and NOT the mpasa grids (as mpasa120 looks OK from Will's plot above). By extension that likely means that mpasa15 and mpasa3p75 are probably OK. But, still not bad to check.

**Comment by wwieder** [comment]
@billsacks mapunits are not regridded correctly on the ne30 and C98 grids
<img width="1062" alt="image" src="https://github.com/user-attachments/assets/dc8d16af-e7c6-4eee-ae9a-5ffced5dc147">

I'll see if I can get enough memory to plot up the higher resolution MPAS grids

**Comment by maritsandstad** [comment]
mapunits seem to have the same problem in the surface data set my end too
![mapunits_output_se_toplevel_map](https://github.com/user-attachments/assets/9b3f4be3-0b51-47b3-b56c-b25f501053b2)

**Comment by mvertens** [comment]
@maritsandstad - thanks! This really helps and points to a problematic part of the mksoitex.F90 routine.

**Comment by billsacks** [comment]
@wwieder @maritsandstad - thank you - this is very helpful! Visually, these look like 0 values. Are you able to quickly confirm that these are indeed exactly 0 over all of Africa - e.g., by setting the range to -1 to 1?

**Comment by wwieder** [comment]
If anyone is interested here's the code needed to look at grids on their native resolution with uxarray

```
import dask
import numpy as np
import pandas as pd
import xarray as xr
import uxarray as ux  #need npl 2024a

import matplotlib
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
from matplotlib.pyplot import cm
import cartopy.crs as ccrs
```

Then point to mesh and data files & plot
```
mesh = '/glade/campaign/cesm/cesmdata/inputdata/share/meshes/ne30pg3_ESMFmesh_cdf5_c20211018.nc'
fin='/glade/derecho/scratch/slevis/temp_work/new_rawdata/tools/mksurfdata_esmf/surfdata_ne30np4.pg3_hist_1850_78pfts_c240905.nc'
ds = ux.open_dataset(mesh, fin)
ds.mapunit.plot()
```

**Comment by wwieder** [comment]
I think you're right @ekluzek.  @dlawrenncar, the mpasa15 grid seems to be OK.  I kind of like how the mapunits plot up, a bit like a quilt, but also checked that PCT_SAND looks OK too
<img width="1102" alt="image" src="https://github.com/user-attachments/assets/0f92eaab-8fbe-47e3-83fe-74bc39880cc8">

**Comment by wwieder** [comment]
@billsacks seems like it (took me a minute to plot a defined range)
<img width="1091" alt="image" src="https://github.com/user-attachments/assets/0873761b-b59b-4ac7-b513-49efccd66250">

**Comment by maritsandstad** [comment]
I get the same, seems like 0
![mapunits_output_se_toplevel_map](https://github.com/user-attachments/assets/e331a3b5-5f8d-4f6e-8eed-730687cd65ee)

**Comment by billsacks** [comment]
Okay, very helpful. @wwieder also confirmed that zbedrock looks fine. I was interested in that one because it uses the same input mesh as soil texture.

@mvertens and I just found something that looks like the culprit... testing now....

**Comment by billsacks** [comment]
The issue @mvertens and I found doesn't solve the issue... though I think is another possible issue that should be fixed, so I'll push that change soon anyway.

**Comment by billsacks** [comment]
As @mvertens and I discussed, I tried changing the implementation of `get_dominant_mapunit` to do something super-simple / wrong: just using the first overlapping source point as the output:

```F90
    if (associated(dynamicMaskList)) then
       do no = 1, size(dynamicMaskList)
          if (size(dynamicMaskList(no)%factor) > 0) then
             dynamicMaskList(no)%dstElement = dynamicMaskList(no)%srcElement(1)
          else
             dynamicMaskList(no)%dstElement = -999
          end if
       end do
    end if
```

This gives reasonable-looking results on the ne30 grid:

![mapunits_src1](https://github.com/user-attachments/assets/41d0477a-a7ff-4ae8-bd54-9d7a88006485)

So there's something specific to the `get_dominant_mapunit` implementation... I'll look deeper.

**Comment by billsacks** [comment]
@mvertens and I have found the issue and the solution (credit to @mvertens for figuring out what needed to be changed). I need to clean up my changes, then will double-check that an ne30 dataset now looks reasonable.

The issue just affects the soil texture variables. **However, it's likely that this has at least a small impact on ALL surface datasets, at all resolutions... we were just (un?)lucky that it doesn't show up more noticeably at other resolutions.** I'll also do a comparison of a before vs. after f09 surface dataset.

**Comment by wwieder** [comment]
This is great.  Thanks @billsacks and @mvertens 
Since we're in the middle of making new surface datasets for CTSM5.3 in #2500 the slight changes that are possible at different resolutions (e.g. f09) don't really matter, as I'd expect we'll make all new datasets with the code changes you're proposing anyway.

**Comment by billsacks** [comment]
Okay, I have tested the fix that I worked on with @mvertens . Maps at ne30 look a lot better now:

mapunit:

![mapunits](https://github.com/user-attachments/assets/92d59e53-e6f4-4179-b31d-3a0a4a83ca53)

Max PCT_SAND in each grid cell:

![PCT_SAND](https://github.com/user-attachments/assets/b0a3be0a-78bc-423a-a486-21630f34918c)


**I just looked back at the plots from @wwieder of the MPAS resolutions. I suspect that these DO exhibit this buggy behavior - it just shows up differently on the MPAS grid: I think the speckling / quilting artifacts that @wwieder attributed to uxarray plotting may in fact be coming from this bug. Regardless, ALL surface datasets should be regenerated with this code fix. PR incoming soon.**

**Comment by billsacks** [comment]
I checked an f09 surface dataset before and after the fix. Only 8 grid cells differ in their `mapunits`, of which 7 differ in the resulting `PCT_SAND` (I haven't checked other variables, but differences should be a subset of these 8 grid cells that differ in their mapunits).

**Comment by billsacks** [comment]
The fix is here: https://github.com/slevis-lmwg/ctsm/pull/9

**Comment by billsacks** [comment]
(I also did a run with the ne30 resolution where I changed the code to set mapunit equal to `mapunit_value_max` everywhere. By doing this, I confirmed that the code for determining `mapunit_value_max` is now working correctly: it was set to 32056 everywhere, as expected.)

**Comment by wwieder** [comment]
I'll be curious if this also addresses #2502?

**Comment by billsacks** [comment]
Yes, good point, this is almost certainly the culprit that explains #2502 - because the nature of this bug means that there would likely be different answers with different processor counts / decompositions.

**Comment by dlawrenncar** [comment]
If this bug is going to affect the MPAS grids as well, can someone work with @briandobbins to create a new ultra high-resolution dataset for the runs that he is working on.  There is some urgency on this, as the runs need to be completed quite soon.  Not urgent as in this needs to be done before the weekend, but early next week preferably.

**Comment by ekluzek** [comment]
@dlawrenncar yes will will do that right away. When we recreate all the datasets the mpasa15 and mpasa3p75 are done at the same time. And rerunning all the datasets will be done by early next week, it may or may not happen over the weekend. So as long as @briandobbins needs one of those we'll get it for him.

**Comment by dmleung** [comment]
@wwieder, I was testing dust in the ne30 grid (#2732) using the same surface dataset yes. The new dust needs some tuning given all the model updates, but thus far I didn't see very significant impacts from the soil variables. The important variables that explicitly affect dust are PCT_CLAY and PCT_NAT_PFT etc. From the conversation here it seems PCT_CLAY is off, so I will need to double check on how this issue impacts my tests. Thanks for letting me know about this!

**Comment by briandobbins** [comment]
In case this is helpful, here's what I see with my 3.75km dataset that I've been using with the ultra-high res CLM50 runs:

ds.mapunits.plot():
![mpasa3p75 plot of mapunits](https://github.com/user-attachments/assets/078f2d82-28db-4a55-a80c-d625b58c4fc3)


ds.mapunits.plot(clim=(-0.1,0.1)):
![dobbins_3p75km_mapunits](https://github.com/user-attachments/assets/6b094745-829a-4776-8449-1b3203a491c9)

**Comment by wwieder** [comment]
@dmleung can you have a look at dust emissions from cases that use this surface dataset:
`/glade/derecho/scratch/slevis/temp_work/new_rawdata/tools/mksurfdata_esmf/surfdata_ne30np4.pg3_hist_1850_78pfts_c240908.nc`


@briandobbins there should also be 3 km mpas grid (`surfdata_mpasa3p75_hist_2000_16pfts_c240908.nc`) you can test out in the same directory

**Comment by briandobbins** [comment]
@wwieder Thanks, that's looking much better to me - here's the clipped new one:
![updated_3 75km_land_file](https://github.com/user-attachments/assets/0cec99a1-e529-4a33-b740-d2eb3a0f672f)

**Comment by wwieder** [comment]
That looks more like planet Earth!

**Comment by ekluzek** [comment]
Hey @jrbuzan I used @wwieder instructions to plot the latest file:

fin = '/glade/campaign/cesm/cesmdata/inputdata/lnd/clm2/surfdata_esmf/ctsm5.3.0/surfdata_ne30np4.pg3_hist_1850_78pfts_c240908.nc'

And got this for the first layer which looks fine to me. Which layer did you plot above?

![bokeh_plot](https://github.com/user-attachments/assets/27f65404-af9e-4e8e-bc65-7a63f5a1d559)

**Comment by billsacks** [comment]
@jrbuzan - yes, this issue would be expected to impact the BULK variable, too.

**Comment by wwieder** [comment]
Just to clarify, we're expecting incorrect BULK on 5.2 and older 5.3 surface datasets created with mksrfdata_esmf- but it should be corrected on the newest 5.3 datasets Sam created (or that @jrbuzan could create with corrections to mksrfdata_esmf that Bill made last week)?

**Comment by billsacks** [comment]
> Just to clarify, we're expecting incorrect BULK on 5.2 and older 5.3 surface datasets created with mksrfdata_esmf- but it should be corrected on the newest 5.3 datasets Sam created (or that @jrbuzan could create with corrections to mksrfdata_esmf that Bill made last week)?

Yes, that's what I'd expect - thanks for stating that more clearly.
